src/V3List.h

```diff
@@ -20,124 +20,324 @@
 #include "config_build.h"
 #include "verilatedos.h"
 
-#include <vector>
+#include "V3Error.h"
 
-//============================================================================
+#include <type_traits>
+#include <utility>
 
-template <class T>
-class V3List;
-template <class T>
-class V3ListEnt;
+//============================================================================
+// The list links (just 2 pointers), to be instantiated as a member in the
+// list element base class 'T_Base'. They are considered mutable, even if the
+// list element is 'const', as they are only used for storing the elements in
+// a V3List. That is, you can store const elements in a V3List.
+template <typename T_Base>
+class V3ListLinks final {
+    // The V3List itself, but nothing else can access the link pointers
+    template <typename B, V3ListLinks<B>& (B::*)(), typename>
+    friend class V3List;
 
-template <class T>
-class V3List final {
-    // List container for linked list of elements of type *T  (T is a pointer type)
-private:
-    // MEMBERS
-    T m_headp = nullptr;  // First element
-    T m_tailp = nullptr;  // Last element
-    friend class V3ListEnt<T>;
+    T_Base* m_nextp = nullptr;  // Next element in list
+    T_Base* m_prevp = nullptr;  // Previous element in list
 
 public:
-    V3List() = default;
-    ~V3List() = default;
-    // METHODS
-    T begin() const { return m_headp; }
-    T end() const { return nullptr; }
-    T rbegin() const { return m_tailp; }
-    T rend() const { return nullptr; }
-    bool empty() const { return m_headp == nullptr; }
-    void reset() {  // clear() without walking the list
-        m_headp = nullptr;
-        m_tailp = nullptr;
+    V3ListLinks() = default;
+    ~V3ListLinks() {
+#ifdef VL_DEBUG
+        m_nextp = reinterpret_cast<T_Base*>(1);
+        m_prevp = reinterpret_cast<T_Base*>(1);
+#endif
     }
+    VL_UNCOPYABLE(V3ListLinks);
+    VL_UNMOVABLE(V3ListLinks);
 };
 
 //============================================================================
+// Generic endogenous (or intrusive) doubly linked list, with links stored
+// inside the elements. The template parameters are:
+// - 'T_Base' is the base type of elements that contains the V3ListLinks
+//   instance as a data member.
+// - 'LinksGetter' is a member function pointer that returns a reference to
+//   the links within 'T_Base'. Note that you should be able to use a data
+//   member pointer, instead of a function pointer, but that is buggy in MSVC.
```

Comment:
MSVC is pants and can't handle member data pointers in various contexts. But with this change it should be good to go now if you hare happy with it.